pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginManager.java

```diff
@@ -111,13 +123,31 @@ public class PluginManager {
         }
       };
 
-  private Map<Plugin, PluginClassLoader> _registry = new HashMap<>();
+  private final ClassWorld classWorld;
+  private final Map<Plugin, PluginClassLoader> _registry;
+
   private String _pluginsDirectories;
   private String _pluginsInclude;
   private boolean _initialized = false;
 
   private PluginManager() {
-    _registry.put(new Plugin(DEFAULT_PLUGIN_NAME), createClassLoader(Collections.emptyList()));
+    if(useLegacyPluginClassloader()) {
+      _registry = new HashMap<>();
+      _registry.put(new Plugin(DEFAULT_PLUGIN_NAME), createClassLoader(Collections.emptyList()));
+      classWorld = null;
+    }
+    else {
+      try {
+        this.classWorld = new ClassWorld();
+        // to simulate behavior of legacy code, however every plugin should have a dedicated realm
+        this.classWorld.newRealm(DEFAULT_PLUGIN_NAME);
+        this.classWorld.newRealm("pinot", ClassLoader.getSystemClassLoader());
+        this._registry = null;
+      } catch (DuplicateRealmException e) {
+        throw new RuntimeException(e);
```

Comment:
I would say the only why this exception can be thrown is if `DEFAULT_PLUGIN_NAME` is equal to `"pinot"`. Am I right?

Reply:
That's correct. Classworld starts like an empty map, but adding a realm comes with a checked exception